### DISCUSSION TITLE
Trying to understand where and how to generate multiple `Command`s from 1 `Event`

### DISCUSSION THREAD

**Discussion by veniamin-ilmer** [hecrj/iced#1999] [Q&A]
I would like to listen to the keyboard CTRL+V event, then depending on what is in the clipboard, evoke **multiple** different `update`s.

I am confused how to do this. I do see both `Subscription` and `Command` have a `batch` method, but I am confused how to create several `Command`s.

Currently inside of the `subscription` method, I use `subscription::events_with`, where after I detect the CTRL+V, my 1 `Message` will be converted into 1 `Command<Message>`.

So a few questions:

1. Where should I read from the clipboard? Inside of the `subscription` method or the `update` method?
2. How do I create multiple `Command`s?

I see there is a `Command::perform` method for example, but I am confused about having to pass in a function into it.

Current code for reference:

    fn subscription(&self) -> iced::Subscription<Message> {
      iced::subscription::events_with(|event, _status| {
        match event {
          iced::Event::Keyboard(iced::keyboard::Event::KeyReleased{modifiers: iced::keyboard::Modifiers::CTRL, key_code: iced::keyboard::KeyCode::V}) => Some(Message::Paste),
          _ => None,
        }
      }
    }

    fn update(&mut self, message: Message) -> Command<Message> {
      match message {
        Message::Paste => {
          return iced::clipboard::read::<Message>(|optional_data| {
            let mut commands = vec![];
            if let Some(data) = optional_data {
              for c in data.chars() {
                if c.is_numeric() {
                  commands.push(Command::/*what goes here?*/(Message::Digit(c)));
                } else {
                  commands.push(Command::/*what goes here?*/(Message::NonDigit(c)));
                }
              }
              return Command::batch(commands);
            }
            Message::Done
          });
        },
        Message::Done => (),
        Message::Digit(c) => { /*handle digit..*/ },
        Message::NonDigit(c) => { /*handle non digit..*/ },
      }
      Command::none()
    }

**Comment by lufte** [comment]
The way I understand it there are no Commands involved here. An iced Command is an asynchronous action that may dispatch a Message as a result or not.

In here, you have a subscription which also may or may not dispatch Messages after each event, so you just need to deal with Messages. As far as I know you can only dispatch one at a time form a subscription, but that should be enough for any use case. For example, let's say you have 2 potential updates that you need to perform on your state depending on what was pasted to the clipboard; you could have 3 Messages to represent all combinations and dispatch only one of them:

1. Message::Thing1WasPasted
2. Message::Thing2WasPasted
3. Message::Things1And2WerePasted

You could also have just one Message with enough internal data for the `update` method to figure out what it needs to do.

Does that make sense?

**Reply by veniamin-ilmer** [reply]
Although your solution will work, it would mean that the window would just be updated once with the final result.

I specifically need to be able to update the window multiple times, as this would be in line with the application I am trying to port over.

For huge amounts of data, you should be able to see the GUI make some really quick updates on the screen.

Oh! I just realized... Could the messages come in out of order? If that's the case, can we force a `view` to be run multiple times during an `update`?

**Reply by lufte** [reply]
Ohh, I see. So it's all to achieve a sort of aesthetic animation of "things happening" or does it serve a functional goal? One way I imagine would work is:
1. Dispatch Message1 from your subscription.
2. `update` processes Message1, runs the first round of updates, and executes a Command which dispatches Message2. Maybe that Command sleeps for a few milliseconds to give the whole animation a perceivable duration.
3. `update` now processes Message2 and runs the second round of updates. Potentially repeat the whole thing more times to achieve the number of steps you want.

It does leave you with a really complex design only to be able to reproduce an animation. I don't know if Iced has support for real animations though.

> can we force a view to be run multiple times during and update?

No I don't think that's possible as it doesn't align with the Elm architecture.

**Reply by veniamin-ilmer** [reply]
Perhaps you can help me out with something else then. If I try and save any data inside of the clipboard reading:

      Message::Paste => {
        return iced::clipboard::read::<Message>(|optional_data| {
          if let Some(data) = optional_data {
            for c in data.chars() {
              if c.is_numeric() {
                self.c = c;
              }
            }
          }
          Message::Done
        });
      }

I get this error:

    error[E0594]: cannot assign to `*self.display`, as it is a captured variable in a `Fn` closure
      --> src\main.rs:97:17
       |
    97 |                 self.c = c;
       |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot assign

    error[E0521]: borrowed data escapes outside of method
       --> src\main.rs:93:16
        |
    80  |     fn update(&mut self, message: Message) -> Command<Message> {
        |               ---------
        |               |
        |               `self` is a reference that is only valid in the method body
        |               let's call the lifetime of this reference `'1`
    ...
    93  |           return iced::clipboard::read::<Message>(|optional_data| {
        |  ________________^
    94  | |           if let Some(data) = optional_data {
    95  | |             for c in data.chars() {
    96  | |               if c.is_numeric() {
    ...   |
    101 | |           Message::Done
    102 | |         });
        | |          ^
        | |          |
        | |__________`self` escapes the method body here
        |            argument requires that `'1` must outlive `'static`

So I am trying to figure out, how am I supposed to store data from the clipboard?

**Reply by lufte** [reply]
Put all the data inside the message (use any of the Enum variants that can hold data within them) and read that data in the `update` method. You won't have access to your state during the asynchronous execution of a Command.

**Reply by veniamin-ilmer** [reply]
I see. Ok, this worked. Thank you for your help.
